### ISSUE TITLE
Jittery movement with TacRP Weapons + PAC3/Wearable Accesories Note

### ISSUE THREAD

**Issue by Lombaxtard** [relaxtakenotes/cool-thirdperson-view#2]
If I move without any weapons or with Half-Life 2 weapons/GMod tools, then I will have a smooth experience using CTV. If I equip a TacRP weapon for example (haven't got resources to test with more weapon bases), then playermodel starts to move erratically, as in some other Thirdperson addons that try to smooth the view by following (with lerp) a player's bone. Again, stock GMod weapons work fine and SWEPs with no custom base will probably do just as good, too. I've got a short video demonstrating this issue, let me know if you want me to share it with you.

PAC3/Wearable Accessories Note: When equipping anything from JMod or binding an object to a bone with PAC3 (a backpack, for example), it will bounce a little when moving which is normal for other TP view addons with smoothing features. The issue however, becomes much more prominent if any of the aforementioned problematic weapons is equipped. Along with PM jitter these objects move around very noticeably.

**Comment by relaxtakenotes** [comment]
As seen in the mods description: Disable frametime or framepace features in your drivers or limit your fps to a static value if you experience any weird jittering or stuttering issues with the camera.

You see, I had this issue as well while making this mod and it disappeared by itself magically. I suppose it was the things I done described above. I also had an algorithm to get a smoother frametime value that's not jittering like an epileptic, but the need for it disappeared... until now I guess. I will cook something up and let you know.

**Comment by relaxtakenotes** [comment]
Download the main branch and test out the cl_nte_frametime_avg_limit and cl_nte_frametime_mode commands. They should help you. Better get down to the root of your problem, though, because for me everything is fine with tacrp, even though I had issues in the past.

**Comment by Lombaxtard** [comment]
> Download the main branch and test out the cl_nte_frametime_avg_limit and cl_nte_frametime_mode commands. They should help you. Better get down to the root of your problem, though, because for me everything is fine with tacrp, even though I had issues in the past.

The new commands haven't helped in the slightest. However, I found the root cause - FPS Stabilizer (https://steamcommunity.com/sharedfiles/filedetails/?id=2913120075). It made PM stutter in TP when holding TacRP, default GMod medkit and some others. Took me around three hours of non-stop testing to narrow it all down. So Shmovment for invisible PM, FPS Stabilizer for jittery PM. Nevertheless, jiggly attachments still stand, I wonder if you could fix that.
P.S. Great work on getting crouching camera fixed, look very sleek now (best used with "Dynamic Height +Crouch Height Fix" to get a higher camera position more akin to PM's actual eye pos. Now all I could hope for is crosshair customization (because yours is actually pretty cool in that it shows if you're actually able to hit what you're aiming at because of camera offset, and camera position customization as well (horizontal and vertical, as distance is already configurable). Really loving your addons, thank you for making and sharing them with the community :3

**Comment by Lombaxtard** [comment]
Sorry for spamming this thread, but would also like to add that it would be great to implement if possible, customization of "bobbing" multiplier for slow walk, run and sprint respectively. If not, then I'll just have to create another thousand of aliases, so not a huge deal breaker .-.

**Comment by relaxtakenotes** [comment]
1) Very sad that it didn't help, because for me it worked just fine.
2) FPS Stabilizer changes a lot of cvars, I suppose one of them messes with frametime or the way calcview executes. Possible suspects would be: cl_cmdrate, cl_updaterate, cl_interp.
3) I'll look into jiggly attachments. It's probably not something I'll be able to fix because of a... peculiar way I calculate stuff.
4) Size, colors, shapes. That enough for crosshair customization?
5) Well I already do variable viewbob speed that changes depending on your max speed. I could allow people to modify certain parts of the formula.

**Comment by relaxtakenotes** [comment]
4. Are we talking about turning it off and on or what?

**Comment by relaxtakenotes** [comment]
Very interesting. It seems these attachments "jiggle" even if you just offset the camera backwards without doing anything special aside from setting calcview to draw the player. This could be a game bug/limitation.

**Comment by relaxtakenotes** [comment]
https://temp.sh/ZLbxm/gmod_1o7meOcDWN.mp4
I don't seem to experience all that much issues with pac/jmod attachments, if any. I don't have anything else installed to mess with the process.

**Comment by Lombaxtard** [comment]
> 3. Ah. Then I probably know why it happens. Might a problem with the order of gathering bone position information on the screen. It might do that before I do all the camera modification stuff and that's why it messes up.
> 
>     4. Are we talking about turning it off and on or what?

4. I didn't mean that, but having an option rather than not having it is probably better..? What I mean was that you could recolor it as well (dark grey to something else)

**Comment by Lombaxtard** [comment]
> https://temp.sh/ZLbxm/gmod_1o7meOcDWN.mp4 I don't seem to experience all that much issues with pac/jmod attachments, if any. I don't have anything else installed to mess with the process.

https://temp.sh/hFaXU/Untitled.wmv

**Comment by relaxtakenotes** [comment]
yeah no i can't replicate it so

**Comment by Lombaxtard** [comment]
> yeah no i can't replicate it so

Alright, will have more tests. By the way, would you be interested in a suggestion for another cl_nte_mode to add? You would've been a pioneer if all works out

**Comment by relaxtakenotes** [comment]
Yea sure, lay the idea out.
I also wanna inform you that I have added: crosshair color/size/border toggle options and a different mechanism for calculating the viewbob speed. You can modify the multiplier for it now. I will push it later together with other additions/fixes I'm eventually going to come up with.
What I won't be adding is crosshair shapes. I'm not looking forward to making a separate material for each shape and using existing solutions in form of crosshair fonts doesn't turn out well.
![gmod_kWMxuJxjkQ](https://github.com/relaxtakenotes/cool-thirdperson-view/assets/91839650/0e2868aa-48a1-4a4b-89d5-48d2f7b0b986)

**Comment by relaxtakenotes** [comment]
https://streamable.com/8wl1nl

I made the viewmodel follow the crosshair position. I need to smooth it out and tweak it so it doesn't flip.

**Comment by Lombaxtard** [comment]
> https://streamable.com/8wl1nl
> 
> I made the viewmodel follow the crosshair position. I need to smooth it out and tweak it so it doesn't flip.

Wow, impressive, didn't think you'd actually get to it
A bit of lerp would be on point for when crosshair snaps from a close to a distant target. Other than that it looks very promising, so nice to see a dream come true :3 Looking a bit into the future here, but I think there's a place for a multiplier that would push the viewmodel closer to the screen when you get up close to a wall, since by turning viewmodel to a side this much could cause visible arms' shoulders that most of the time have no texture on the inside
Can't stress enough how much this is exciting for me

**Comment by relaxtakenotes** [comment]
Ah damn, you saw the deleted thing. I wanted to show you the end result instead.
About looking into the future - you're right. I will need to do something about that because it sometimes looks like this.
![gmod_axYRc5BzkI](https://github.com/relaxtakenotes/cool-thirdperson-view/assets/91839650/d26f6165-fb7a-45ac-ac7a-f474031b9e3d)

**Comment by relaxtakenotes** [comment]
You can try everything I did in this branch: https://github.com/relaxtakenotes/cool-thirdperson-view/tree/dev
iz pretty ruff... So expect jank.

**Comment by Lombaxtard** [comment]
In case it'll be helpful: [thirdperson_effect_dev] addons/thirdperson_effect_dev/lua/autorun/client/nice_thirdperson_effect.lua:341: attempt to index upvalue 'wish_pos' (a nil value)
  1. fn - addons/thirdperson_effect_dev/lua/autorun/client/nice_thirdperson_effect.lua:341
   2. unknown - lua/ulib/shared/hook.lua:109 (x294)

I'll try it on a cleaner addonlist

**Comment by relaxtakenotes** [comment]
I pushed it to workshop with calcview priority system enabled. Try it out

**Comment by Lombaxtard** [comment]
I did, and it's awesome! Nonetheless, some issues I've encountered: only map's collision triggers anti-clipping system, not sure if including props would've affected performance; enabling hybrid firstperson forces mode 1, I think, because I have to re-exec my preset for CTV to get back to thirdperson; not sure what can be done about it, probably just ignore it, but when "hidden" crosshair is triggered maybe there'd better be some viewmodel clues? Right now if you peek from behind a wall (depends on the left/right side) then you will be having your crosshair on a wall, actual crosshair behind a wall, and able to essentially blindfire. Still, that's probably just collateral given the circumstances, I'm still very impressed how well your execution of this firstperson concept works; still looking forward to camera offset control (X/Z) :)

**Comment by relaxtakenotes** [comment]
Yea, hybrid firstperson does that. Once you turn it off it'll come back to normal.
I forgot about the camera offsets. They'll be the next thing I do.

**Comment by relaxtakenotes** [comment]
And regarding anti clipping system I guess I need to fiddle around with trace parameters.

**Comment by relaxtakenotes** [comment]
Fixed that clipping stuff, added thirdperson offsets, made hybrid firstperson not force you in firstperson

**Comment by relaxtakenotes** [comment]
> Remember to credit me for each :Set :Mul :Adds that you include

only after u credit me for the fix after that one gmod update

**Comment by relaxtakenotes** [comment]
> > > Remember to credit me for each :Set :Mul :Adds that you include
> > 
> > 
> > only after u credit me for the fix after that one gmod update
> 
> Wut?

yea u know what im talking about, earthykiller...

**Comment by relaxtakenotes** [comment]
yep cause ur lurking my shit like crazy

**Comment by relaxtakenotes** [comment]
?

**Comment by Lombaxtard** [comment]
So, managed to test it with just CTV, PAC3, CalcView Priority System, EFT Equipment Megapack for backpack from that vid in the tunnels
I went ahead and put my back against the wall to make PM half-visible and attached the same bp to Spine 2. Near-wall fading feature doesn't affect stuff like PAC3 so i started moving with CTV and then with native GMod thirdperson. With CTV I can see backpack going up and down as I step, with it disabled it stays bound to the bone I've asigned it to without extra movements. Only logical thing to me that could've caused this was CTV's viewbob, so I disabled it. However, that did not help. Maybe I should share my saved preset with backpack for you to see it. But like I said, View Extension TP and regular GMod TP don't cause this

**Comment by Lombaxtard** [comment]
By the way, worth noting that, I guess, most CalcView stuff will break hybrid mode by putting viewmodel somewhere to the side and it moves like on the screen like it's has worldmodel's weapon position, if that makes sense (tested with CBobbing). Also, enabling hybrid hides viewmodel if not in cl_nte_enabled 1. Also, I think it would've been better to limit immersive firstperson nearz not only when looking down, but when looking up as well, just by a couple degrees. Since VM still flips a little bit too much for it too look right. And before I forgot, using scoped weapons while in immersive firstperson mode (tested with hybrid enabled) denies magnification entirely. Still testing with TacRP by the way

**Comment by Lombaxtard** [comment]
One more thing I Did forget is that if I disable thirdperson, move away (any distance), then enable CTV back on the camera seems to transition from the "last enabled" spot to where I am now. Not in real time, of course. Plenty fast, but still very noticeable

**Comment by relaxtakenotes** [comment]
> One more thing I Did forget is that if I disable thirdperson, move away (any distance), then enable CTV back on the camera seems to transition from the "last enabled" spot to where I am now. Not in real time, of course. Plenty fast, but still very noticeable

Fixed

> By the way, worth noting that, I guess, most CalcView stuff will break hybrid mode by putting viewmodel somewhere to the side and it move on the screen like it's has worldmodel's weapon position, if that makes sense (tested with CBobbing). Moreover, enabling hybrid hides viewmodel entirely if not in cl_nte_enabled 1. Also, I think it would've been better to limit immersive firstperson nearz not only when looking down, but when looking up as well, just by a couple degrees. Since VM still flips a little bit too much for it too look right. And before I forgot, using scoped weapons while in immersive firstperson mode (tested with hybrid enabled) denies magnification entirely. Still testing with TacRP by the way

Made the limit stricter. Made it respect your fov as defined from the fov_desired cvar. Won't do anything about conflicting mods tho.

> So, managed to test it with just CTV, PAC3, CalcView Priority System, EFT Equipment Megapack for backpack from that vid in the tunnels I went ahead and put my back against the wall to make PM half-visible and attached the same bp to Spine 2. Near-wall fading feature doesn't affect stuff like PAC3 so i started moving with CTV and then with native GMod thirdperson. With CTV I can see backpack going up and down as I step, with it disabled it stays bound to the bone I've asigned it to without extra movements. Only logical thing to me that could've caused this was CTV's viewbob, so I disabled it. However, that did not help. Maybe I should share my saved preset with backpack for you to see it. But like I said, View Extension TP and regular GMod TP don't cause this

I'll try to replicate it this way, thanks.

**Comment by Lombaxtard** [comment]
After last update I noticed something weird...
https://temp.sh/CULCJ/hidden.wmv
Actually, due to me being engaged with testing stuff I haven't got my hands on the crosshaie until recently. But just as I was contemplating editing it I loaded into a map and This hapenned.. Also, not only settings for crosshair wouldn't apply, "hidden" crosshair completely stopped drawing (last moments switching shoulders near a wall)

**Comment by relaxtakenotes** [comment]
> After the last update I noticed something weird... https://temp.sh/CULCJ/hidden.wmv Actually, due to me being engaged with testing stuff I haven't got my hands on the crosshair until recently. But just as I was contemplating editing it I loaded into a map and This hapenned.. Also, not only settings for crosshair wouldn't apply, "hidden" crosshair completely stopped drawing (last moments switching shoulders near a wall)
> 
> P.S. You forgot to remove now redundant "Install CalcView Priority System for it to work" from GitHub's README .-.

Looks like the "hidden" trace starts from the 0,0,0 point in the world, or from where you spawned. I'll look into it

Deleted the calcview thing from the readme lol

**Comment by Lombaxtard** [comment]
> 

So crosshair not changing colors is a new conflict I have to find, right? Or did it somehow get broken and you might have an idea why that might be?

**Comment by relaxtakenotes** [comment]
> > 
> 
> So crosshair not changing colors is a new conflict I have to find, right? Or did it somehow get broken and you might have an idea why that might be?

Don't know at the moment.

**Comment by relaxtakenotes** [comment]
Oh yeah, seems to work fine for me. Try getting the master version from the repo and typing out "developer 1" in console. Check for a green line. If it doesn't come out of your view, then it's getting messed with. If you can't see it or can see it just barely, then everything is fine.

**Comment by Lombaxtard** [comment]
> Oh yeah, seems to work fine for me. Try getting the master version from the repo and typing out "developer 1" in console. Check for a green line. If it doesn't come out of your view, then it's getting messed with. If you can't see it or can see it just barely, then everything is fine.

I tried this version. Green line flickers aggressively, but now hidden crosshair behaves properly and both types of crosshair are able to recolor

**Comment by relaxtakenotes** [comment]
> Will do.
> 
> By the way, every Immersive Firstperson I know of has this issue: https://temp.sh/GnaSf/flickering.wmv Not sure it can be fixed

Looks like changing the clipping plane is a bad idea! It will be reverted in the next update. Long story short it's because of float precision.

> > Oh yeah, seems to work fine for me. Try getting the master version from the repo and typing out "developer 1" in console. Check for a green line. If it doesn't come out of your view, then it's getting messed with. If you can't see it or can see it just barely, then everything is fine.
> 
> I tried this version. Green line flickers aggressively, but now hidden crosshair behaves properly and both types of crosshair are able to recolor

Oh, nice.

**Comment by Lombaxtard** [comment]
Check This out: https://temp.sh/ejupr/now%20what.wmv
I'll try the newest version now, just in case

**Comment by relaxtakenotes** [comment]
> Check This out: https://temp.sh/ejupr/now%20what.wmv I'll try the newest version now, just in case

What is this...

**Comment by Lombaxtard** [comment]
> > Check This out: https://temp.sh/ejupr/now%20what.wmv I'll try the newest version now, just in case
> 
> What is this...

What makes ot a little funnier from the dev perspective that I haven't added a single addon since

**Comment by relaxtakenotes** [comment]
> > > Check This out: https://temp.sh/ejupr/now%20what.wmv I'll try the newest version now, just in case
> > 
> > 
> > What is this...
> 
> What makes ot a little funnier from the dev perspective that I haven't added a single addon since

There's nothing in the video. I don't understand what you're getting at, unless my mod teleported your view into oblivion and you can't see shit.

**Comment by Lombaxtard** [comment]
Fixed..? I dunno, maybe it was a map problem, or my game starts to become a schizo

Here's that vid: https://temp.sh/bMapG/now_what.wmv

**Comment by relaxtakenotes** [comment]
> Fixed..? I dunno, maybe it was a map problem, or my game starts to become a schizo
> 
> Here's that vid: https://temp.sh/bMapG/now_what.wmv

Oh yeah, I fixed that in the last commit. Apparently my check for the calcview being modified by other mods was a bit too rough, so I disabled it.

**Comment by Lombaxtard** [comment]
> > Fixed..? I dunno, maybe it was a map problem, or my game starts to become a schizo
> > Here's that vid: https://temp.sh/bMapG/now_what.wmv
> 
> Oh yeah, I fixed that in the last commit. Apparently my check for the calcview being modified by other mods was a bit too rough, so I disabled it.

Good to know... 

Btw, I reckon you weren't able to reproduce that?


> So, managed to test it with just CTV, PAC3, CalcView Priority System, EFT Equipment Megapack for backpack from that vid in the tunnels I went ahead and put my back against the wall to make PM half-visible and attached the same bp to Spine 2. Near-wall fading feature doesn't affect stuff like PAC3 so i started moving with CTV and then with native GMod thirdperson. With CTV I can see backpack going up and down as I step, with it disabled it stays bound to the bone I've asigned it to without extra movements. Only logical thing to me that could've caused this was CTV's viewbob, so I disabled it. However, that did not help. Maybe I should share my saved preset with backpack for you to see it. But like I said, View Extension TP and regular GMod TP don't cause this

**Comment by relaxtakenotes** [comment]
> > > Fixed..? I dunno, maybe it was a map problem, or my game starts to become a schizo
> > > Here's that vid: https://temp.sh/bMapG/now_what.wmv
> > 
> > 
> > Oh yeah, I fixed that in the last commit. Apparently my check for the calcview being modified by other mods was a bit too rough, so I disabled it.
> 
> Good to know...
> 
> Btw, I reckon you weren't able to reproduce that?
> 
> > So, managed to test it with just CTV, PAC3, CalcView Priority System, EFT Equipment Megapack for backpack from that vid in the tunnels I went ahead and put my back against the wall to make PM half-visible and attached the same bp to Spine 2. Near-wall fading feature doesn't affect stuff like PAC3 so i started moving with CTV and then with native GMod thirdperson. With CTV I can see backpack going up and down as I step, with it disabled it stays bound to the bone I've asigned it to without extra movements. Only logical thing to me that could've caused this was CTV's viewbob, so I disabled it. However, that did not help. Maybe I should share my saved preset with backpack for you to see it. But like I said, View Extension TP and regular GMod TP don't cause this

yep... yet.

**Comment by Lombaxtard** [comment]
And you probably never would have. I found tge culprit, it was FPS. No, not the fluctuating framerate, but an FPS lock below average. But like I said, an issue only visible in CTV. Try lowest fps_max 30 just to be sure and get a backpack.You probably won't need to pay much attention to see how it is. Then you can get the Camera tool and spawn it with RMB, then walk in circles around it to confirm that nothing odd is hapenning.
I don't know if that can be fixed, but it sure a hell would've been nice

**Comment by relaxtakenotes** [comment]
> And you probably never would have. I found the culprit, it was FPS. No, not the fluctuating framerate, but an FPS lock below average. But like I said, an issue only visible in CTV. Try lowest fps_max 30 just to be sure and get a backpack. You probably won't need to pay much attention to see how it is. Then you can get the Camera tool and spawn it with RMB, then walk in circles around it to confirm that nothing odd is hapenning. I don't know if that can be fixed, but it sure a hell would've been nice

wow cringe... i knew low fps caused issues in some aspects of this mod but i didn't think it'd be this

**Comment by Lombaxtard** [comment]
Latest version somehow has this when in mode 1 with hybrid enabled. Any weapon (tested with ARC9 too), vm always facing upwards
![image](https://github.com/relaxtakenotes/cool-thirdperson-view/assets/75931669/83a146cb-2790-43e1-9414-e62f29898048)
I guess I missed it because I was focused on TP and crosshairs during previous testing

**Comment by Lombaxtard** [comment]
Oh, and a quick little suggestion to make it compatible with ARC9 leaning when in CTV's hybrid mode. It works well, but viewmodel gets pushed to sides and that's the only real problem. Well, hopefully it wouldn't be impossible..

**Comment by Lombaxtard** [comment]
sendhelp pls
https://temp.sh/tMMwT/imhihgh.wmv

**Comment by relaxtakenotes** [comment]
Okay, back to the grind. I was taking a break from this stuff.
First of all I made a new feature that modifies your bullets to come out of your actual gun and not your eyes: https://youtu.be/IVFxEmvyvB8
Second, in my development version I can not replicate your issue with the viewmodels aiming up.

Pushed am update to the repo. Still has some bugs, but you can try it out.

**Comment by Lombaxtard** [comment]
> Okay, back to the grind. I was taking a break from this stuff. First of all I made a new feature that modifies your bullets to come out of your actual gun and not your eyes: https://youtu.be/IVFxEmvyvB8 Second, in my development version I can not replicate your issue with the viewmodels aiming up.
> 
> Pushed am update to the repo. Still has some bugs, but you can try it out.

I started to think my nagging finally got you

New feature is actually pretty awesome, but the first thing that comes to mind after watching your vid is whether you'de be able to do shooting through thin walls and props or not

About viewmodel issue.. I can't reliably reproduce this, last video I've sent (if you watched it before it was removed from temp.sh) was captured while only having CTV, ARC9 Base + several ARC9 weapon packs only from EFT series and TacMove. But I got this error through testing before installing TacMove and having jsut the CTV and ARC9. I was also able to get this bug with TacRP, as seen in a screenshot above. It seemed to me that the bigger the freeze from spawning the weapon (I've got a pretty slow HDD), the more chances that it will turn out like that. I've sent you a video before with ARC9 weapon held showing "clipping plane) issue whule in hybrid mode and the weapon was fine. 
I guess I'll try this build with some hope, but that vm crap really looks like it was introduced in an update at some point.

By the way, I had some time to play around with thirdperson offsets and was disappointed that it took me so long to actually get to testing it and find out that it doesn't exactly work as expected. I'll get a video and send it here sometime soon

**Comment by relaxtakenotes** [comment]
> > Okay, back to the grind. I was taking a break from this stuff. First of all I made a new feature that modifies your bullets to come out of your actual gun and not your eyes: https://youtu.be/IVFxEmvyvB8 Second, in my development version I can not replicate your issue with the viewmodels aiming up.
> > Pushed am update to the repo. Still has some bugs, but you can try it out.
> 
> I started to think my nagging finally got you
> 
> New feature is actually pretty awesome, but the first thing that comes to mind after watching your vid is whether you'de be able to do shooting through thin walls and props or not
> 
> About viewmodel issue.. I can't reliably reproduce this, last video I've sent (if you watched it before it was removed from temp.sh) was captured while only having CTV, ARC9 Base + several ARC9 weapon packs only from EFT series and TacMove. But I got this error through testing before installing TacMove and having only the CTV and ARC9. I was also able to get this bug with TacRP, as seen in a screenshot above. It seemed to me that the bigger the freeze from spawning the weapon (I've got a pretty slow HDD), the more chances that it will turn out like that. I've sent you a video before with ARC9 weapon held showing "clipping plane) issue while in hybrid mode and the weapon was fine. I guess I'll try this build with some hope, but that vm crap really looks like it was introduced in an update at some point.
> 
> By the way, I had some time to play around with thirdperson offsets and was disappointed that it took me so long to actually get to testing it and find out that it doesn't exactly work as expected. I'll get a video and send it here sometime soon

I'll make it do some checks so you can't shoot through thin walls.
I get what you mean by the thirdperson offsets not being correct. I forgot about certain math stuff that I also need to do.
Also you can do this tomfoolery with the new feature: https://temp.sh/OdtVh/gmod_J83VStYcXC.mp4, the range is limited to 72 hammer units, beyond which your new shooting position will be forced to eye position. Very abusable, especially by cheaters, because I'm letting the client set the hand position and viewmodel radius, with which you can reach people from ridiculous places.

**Comment by Lombaxtard** [comment]
I guess it's too late
https://temp.sh/HvDQO/longtimenohear.wmv

> I get what you mean by the thirdperson offsets not being correct.

**Comment by Lombaxtard** [comment]
> Also you can do this tomfoolery with the new feature

Looks fun, pretty sure you can do some cool stuff with this
Not really into multiplayer in GMod, so I mostly do singleplayer shenanigans or a bit of 1v1 deathmatch/play pretend. I some really basic stuff may be implemented to prevent outright bs in multiplayer. Or maybe it's GMod and everything should be as accessible as possible

**Comment by relaxtakenotes** [comment]
> > Also you can do this tomfoolery with the new feature
> 
> Looks fun, pretty sure you can do some cool stuff with this Not really into multiplayer in GMod, so I mostly do singleplayer shenanigans or a bit of 1v1 deathmatch/play pretend. I some really basic stuff may be implemented to prevent outright bs in multiplayer. Or maybe it's GMod and everything should be as accessible as possible

yep! https://cdn.discordapp.com/attachments/999798596750811136/1133479414936649919/gmod_7pzoqPQsPx.mp4

**Comment by Lombaxtard** [comment]
> > > Also you can do this tomfoolery with the new feature
> > 
> > 
> > Looks fun, pretty sure you can do some cool stuff with this Not really into multiplayer in GMod, so I mostly do singleplayer shenanigans or a bit of 1v1 deathmatch/play pretend. I some really basic stuff may be implemented to prevent outright bs in multiplayer. Or maybe it's GMod and everything should be as accessible as possible
> 
> yep! https://cdn.discordapp.com/attachments/999798596750811136/1133479414936649919/gmod_7pzoqPQsPx.mp4

Almost the leaning fix I've mentioned before, reminds me of blindfire shooting feature from TacRP

**Comment by Lombaxtard** [comment]
I know I'm beating a death horse at this point, but I've got a better demonstration of the issue
It's funny, cause the cause -.- is FPS again (or FPS drops rather). Try setting fps_max to 1 for a couple of seconds while in hybrid fp and you should get the same viewmodel bug - I've just tested exactly that to be damn sure
https://temp.sh/qAZVL/titled1.wmv

**Comment by relaxtakenotes** [comment]
> I know I'm beating a death horse at this point, but I've got a better demonstration of the issue It's funny, cause the cause -.- is FPS again (or FPS drops rather). Try setting fps_max to 1 for a couple of seconds while in hybrid fp and you should get the same viewmodel bug - I've just tested exactly that to be damn sure https://temp.sh/qAZVL/titled1.wmv

weird how a lag spike can cause this. i'll look into this at some point

**Comment by Lombaxtard** [comment]
Ok, great
Let me know if and when you'll get to fixing thirdperson offsets
Would also love to see a "platformer mode" in CTV (Ratchet&Clank 1, 2, 3 style to be precise), because GMod's default thirdperson_platformer is kind of cancer and there's really no good alternative to it (View Extension's directional movement is extremely janky and doesn't seem to utilize native way of handling it as this "thirdperson_platformer" mode does
I guess a toggle for aiming mode would also be nice? There's plenty of potential with custom configs and auto-zoom when holding RMB is going to mess up with some them it at some point (actually thinking about ways of messing with offsets (if they're gonna be patched) right now)

**Comment by relaxtakenotes** [comment]
> Ok, great Let me know if and when you'll get to fixing thirdperson offsets Would also love to see a "platformer mode" in CTV (Ratchet&Clank 1, 2, 3 style to be precise), because GMod's default thirdperson_platformer is kind of cancer and there's really no good alternative to it (View Extension's directional movement is extremely janky and doesn't seem to utilize native way of handling it as this "thirdperson_platformer" mode does I guess a toggle for aiming mode would also be nice? There's plenty of potential with custom configs and auto-zoom when holding RMB is going to mess up with some them it at some point (actually thinking about ways of messing with offsets (if they're gonna be patched) right now)

pretty sure i fixed the offset issue already, but i pushed the other most recent changes just now

**Comment by Lombaxtard** [comment]
> pretty sure i fixed the offset issue already

Indeed they are

Btw, managed to get these when in default GMod thirdperson with CTV set 0

1:
[thirdperson_effect] addons/thirdperson_effect/lua/autorun/client/aaaaa_nice_thirdperson_effect.lua:519: attempt to call global 'invalidate_vm_data' (a nil value)
    1. fn - addons/thirdperson_effect/lua/autorun/client/aaaaa_nice_thirdperson_effect.lua:519
        2. unknown - lua/ulib/shared/hook.lua:109
            3. DrawModel - [C]:-1
                4. unknown - lua/weapons/mg_base/modules/client/cl_worldmodel_render.lua:16
                
2:
[thirdperson_effect] addons/thirdperson_effect/lua/autorun/client/aaaaa_nice_thirdperson_effect.lua:519: attempt to call global 'invalidate_vm_data' (a nil value)
    1. fn - addons/thirdperson_effect/lua/autorun/client/aaaaa_nice_thirdperson_effect.lua:519
        2. unknown - lua/ulib/shared/hook.lua:109

**Comment by relaxtakenotes** [comment]
> > pretty sure i fixed the offset issue already
> 
> Indeed they are
> 
> Btw, managed to get these when in default GMod thirdperson with CTV set 0
> 
> 1: [thirdperson_effect] addons/thirdperson_effect/lua/autorun/client/aaaaa_nice_thirdperson_effect.lua:519: attempt to call global 'invalidate_vm_data' (a nil value) 1. fn - addons/thirdperson_effect/lua/autorun/client/aaaaa_nice_thirdperson_effect.lua:519 2. unknown - lua/ulib/shared/hook.lua:109 3. DrawModel - [C]:-1 4. unknown - lua/weapons/mg_base/modules/client/cl_worldmodel_render.lua:16
> 
> 2: [thirdperson_effect] addons/thirdperson_effect/lua/autorun/client/aaaaa_nice_thirdperson_effect.lua:519: attempt to call global 'invalidate_vm_data' (a nil value) 1. fn - addons/thirdperson_effect/lua/autorun/client/aaaaa_nice_thirdperson_effect.lua:519 2. unknown - lua/ulib/shared/hook.lua:109

fixed

**Comment by Lombaxtard** [comment]
lua/autorun/client/aaaaa_nice_thirdperson_effect.lua:545: attempt to compare nil with number
    1. fn - addons/(!loose) ctv - cool thirdperson&firstperson view/lua/autorun/client/aaaaa_nice_thirdperson_effect.lua:545
        2. unknown - lua/ulib/shared/hook.lua:109

**Comment by relaxtakenotes** [comment]
> 

pushed an untested fix

**Comment by Lombaxtard** [comment]
![image](https://github.com/relaxtakenotes/cool-thirdperson-view/assets/75931669/ce0d2410-4c7f-4c21-bc6a-6ad936b1015e)

**Comment by Lombaxtard** [comment]
Still there

**Comment by relaxtakenotes** [comment]
> The camera resetting to something each time you stop moving. Only noticeable if you move a camera at that specific moment (which is most of the time, actually) https://temp.sh/NQOYO/thedeeperwego.wmv

wow awesome, this will be fixed

**Comment by Lombaxtard** [comment]
Could this possibly be fixed? Not a big deal, but still some neat stuff (smooth offset transition while moving, snappy while standing still)
https://temp.sh/hHkGJ/snappysmoothness.wmv

**Comment by relaxtakenotes** [comment]
> Could this possibly be fixed? Not a big deal, but still some neat stuff (smooth offset transition while moving, snappy while standing still) https://temp.sh/hHkGJ/snappysmoothness.wmv

yes

**Comment by Lombaxtard** [comment]
I want my updates and fixes 2 hours ago! Dear God, this developer is so lazy I'm surprised people are even active in this thread........
Before I forgot, "cl_nte_fov_thing_was_reset_lol_epic" IS a variable just for giggles or does it revert stuff back to the old "FOV ways" when set to 0? I'm also wondering if you just forgot it exists lol

**Comment by relaxtakenotes** [comment]
> I want my updates and fixes 2 hours ago! Dear God, this developer is so lazy I'm surprised people are even active in this thread........ Before I forgot, "cl_nte_fov_thing_was_reset_lol_epic" IS a variable just for giggles or does it revert stuff back to the old "FOV ways" when set to 0? I'm also wondering if you just forgot it exists lol

it's for backwards compatibility with older configs, since i had to change how the fov settings work

**Comment by Lombaxtard** [comment]
2 points:
1. Have you considered a possibility to apply fading effect to PAC3 attachments?
2. Another nice (toggleable) feature besides platformer mode would be a crosshair that always gravitates towards the center of the screen, just like in TFA Thirdperson and Enhanced Thirdperson

Enhanced Thirdperson: ![Screenshot (1327)](https://github.com/relaxtakenotes/cool-thirdperson-view/assets/75931669/69f78a84-66ac-4b68-be24-3070f65a976e)
CTV: ![Screenshot (1328)](https://github.com/relaxtakenotes/cool-thirdperson-view/assets/75931669/e070584b-4847-4275-9e5e-609b23633278)

**Comment by relaxtakenotes** [comment]
> 2 points:
> 
>     1. Have you considered a possibility to apply fading effect to PAC3 attachments?
> 
>     2. Another nice (toggleable) feature besides platformer mode would be a crosshair that always gravitates towards the center of the screen, just like in TFA Thirdperson and Enhanced Thirdperson
> 
> 
> Enhanced Thirdperson: ![image](https://user-images.githubusercontent.com/75931669/257870689-b111f2a6-dbf8-4e38-8ba4-a5fd80ddcd32.png)
> 
> CTV: ![image](https://user-images.githubusercontent.com/75931669/257870834-1453c694-71cc-4a7a-afbb-262bc8260b82.png)

1) i have. i also have ideas on how to do it
2) i'm afraid i'd have to do an unholy amounts of fuckery with movement which will cause issues with certain mods. might do it at some point th

**Comment by Lombaxtard** [comment]
[(!loose) ctv - cool thirdperson&firstperson view] addons/(!loose) ctv - cool thirdperson&firstperson view/lua/autorun/client/aaaaa_nice_thirdperson_effect.lua:545: attempt to compare nil with number
    1. fn - addons/(!loose) ctv - cool thirdperson&firstperson view/lua/autorun/client/aaaaa_nice_thirdperson_effect.lua:545
        2. unknown - lua/ulib/shared/hook.lua:109
        
Got a little better understanding of why this lua error happens
When I first boot into a map I have an empty loadout (literally no weapons, not even empty hands or a holster) this error comes reliably up each time. After I spawn any weapon, lua crap goes away. However, if I force-remove all weapon slots I may be holding, then the error doesn't come back : |

Actually, I was able to get it one time during gameplay while holding TacRP as usual, but it was a one off thing and I wasn't able to get it again while playing normally again

**Comment by Lombaxtard** [comment]
[(!loose) ctv - cool thirdperson&firstperson view] addons/(!loose) ctv - cool thirdperson&firstperson view/lua/autorun/client/aaaaa_nice_thirdperson_effect.lua:522: bad argument #1 to 'WriteFloat' (number expected, got nil)
  1. WriteFloat - [C]:-1
   2. callback - addons/(!loose) ctv - cool thirdperson&firstperson view/lua/autorun/client/aaaaa_nice_thirdperson_effect.lua:522
    3. unknown - lua/includes/modules/cvars.lua:48
    
Another one

**Comment by Lombaxtard** [comment]
lua/autorun/client/am_music.lua:39: attempt to index global 'am_current_song' (a nil value)
    1. callback - lua/autorun/client/am_music.lua:39
        2. unknown - lua/includes/modules/cvars.lua:48

Sorry for posting this here, I'm just not sure if this error is worth a whole new thread on the AM github page

**Comment by relaxtakenotes** [comment]
action music is spaghetti idk if i'll maintain it, i only like the tool i made for it

**Comment by Lombaxtard** [comment]
> action music is spaghetti idk if i'll maintain it, i only like the tool i made for it

But I like spaghetti... So much for giving Nombat hard time lol
Funny thing, but I downloaded it again and try to launch with no luck, simply got stuck on "installing ffmpeg"

**Comment by relaxtakenotes** [comment]
> > action music is spaghetti idk if i'll maintain it, i only like the tool i made for it
> 
> But I like spaghetti... So much for giving Nombat hard time lol Funny thing, but I downloaded it again and try to launch with no luck, simply got stuck on "installing ffmpeg"

it's still better than nombat in terms of features and making packs for it.
install ffmpeg manually and put it either in the same directory as the app or put it in path

**Comment by Lombaxtard** [comment]
Some more fun stuff in case you decide to come back

![image](https://github.com/relaxtakenotes/cool-thirdperson-view/assets/75931669/e4fbb2ba-c45c-4fb7-bca8-dc88dd0fc90d)

**Comment by Lombaxtard** [comment]
https://oshi.at/Anfdf/telecrouch.wmv

Btw, temprorarily fixed viewmodel getting broken by disabling TraceLine for vm in the code .-.

**Comment by relaxtakenotes** [comment]
did a lot of cleaning in the code and perhaps fixed some of your issues in the process

**Comment by Lombaxtard** [comment]
> did a lot of cleaning in the code and perhaps fixed some of your issues in the process

I can see that, I'm impressed by how many things are actually fixed now. Sorry for pestering you with requests, guess that's what made you burn out..?

Anyway, what was got rid of:
1. Hybrid firstperson acting up after a moment of low fps/spawnfreeze (no amount of "fps_max 1" can make it break now, which is fn awesome)
2. Instant transitions when standing still (remember that vanity fix I asked for, for when transitions are smooth if you're moving but snappy if you're not? Now it's ALL smooth and silky)
3. Telecrouching with ARC9 leaning (not 100% sure though, wasn't able to test in the same spot yet)
4. Thirdperson offsets making camera position snap if not on "default_side -1" when stopping (not particularly sure, but seems to be the case pretty much)

What's still present: 
1. Spawning/being with "null" weapon or with no weapon at all: addons/!ctv 9,23,2023/lua/autorun/client/!!!!!nice_thirdperson_effect.lua:504: attempt to compare userdata with number
  1. fn - addons/!ctv 9,23,2023/lua/autorun/client/!!!!!nice_thirdperson_effect.lua:504
   2. unknown - lua/ulib/shared/hook.lua:109 (x4)
all until I strip "null" or whatever it is, I'm a bit guessing here. Or simply equip a weapon AFTER this error has occurred - it seems that if I get "normal" holdtype after holding some guns it will still appear until I switch back to 'em/try to strip "nothing", then it will fix itself for the next map session


New stuff:
1. When holding ARC9 weapons: addons/!ctv 9,23,2023/lua/autorun/client/!!!!!nice_thirdperson_effect.lua:504: attempt to compare userdata with number
  1. fn - addons/!ctv 9,23,2023/lua/autorun/client/!!!!!nice_thirdperson_effect.lua:504
   5. unknown - lua/ulib/shared/hook.lua:109 (x1953)
So far so good..?

**Comment by relaxtakenotes** [comment]
no i just have random phases when i want to play gmod and when i dont

i fixed the bug with arc9 weapons already before you responded, but i guess after you got it.

**Comment by Lombaxtard** [comment]
> no i just have random phases when i want to play gmod and when i dont
> 
> i fixed the bug with arc9 weapons already before you responded, but i guess after you got it.

Hmm, guess so
I was thinking for a while now, think you can add lerp to vm in hybrid firstperson?

**Comment by relaxtakenotes** [comment]
it already lerps tho?

**Comment by relaxtakenotes** [comment]
![image](https://github.com/relaxtakenotes/cool-thirdperson-view/assets/91839650/58376b7b-73c2-4cf1-a997-8f11e019d434)

**Comment by Lombaxtard** [comment]
> ![image](https://user-images.githubusercontent.com/91839650/270109946-58376b7b-73c2-4cf1-a997-8f11e019d434.png)

Is this a new (september) addition? It looks marginally better now, I just assumed it was left unchanged and was still extra-snappy when moving mouse over objects that are close

**Comment by Lombaxtard** [comment]
Btw, I saw "experimental firstperson" in Draconic Base recently
Definitely not as polished as what CTV is and how it has improved even more, but I noticed a neat feature of having body visibility while also having visible viewmodel
It's not great, but what if you actually clip playermodel? It doesn't really make sense for Draconic, but for CTV there's a camera angle restriction which wouldn't allow you to look into your clipped guts
Just saying

**Comment by relaxtakenotes** [comment]
> > ![image](https://user-images.githubusercontent.com/91839650/270109946-58376b7b-73c2-4cf1-a997-8f11e019d434.png)
> 
> Is this a new (september) addition? It looks marginally better now, I just assumed it was left unchanged and was still extra-snappy when moving mouse over objects that are close

well it was there ever since i added the feature so...



> Btw, I saw "experimental firstperson" in Draconic Base recently Definitely not as polished as what CTV is and how it has improved even more, but I noticed a neat feature of having body visibility while also having visible viewmodel It's not great, but what if you actually clip playermodel? It doesn't really make sense for Draconic, but for CTV there's a camera angle restriction which wouldn't allow you to look into your clipped guts Just saying

idk how to render the playermodel and the viewmodel at the same time. he probably rebuilt the playermodel from the ground up like gmod legs 3 did for its legs

**Comment by relaxtakenotes** [comment]
oh yes im definitely not going to do this shit 
![image](https://github.com/relaxtakenotes/cool-thirdperson-view/assets/91839650/77fb5500-24d5-4f56-a5ff-ca0a196fa3ea)

**Comment by relaxtakenotes** [comment]
what i can do is remove the hooks that alter the view and only keep the hook that renders the model... muhahaha
https://i.rlxx.ru/I3b5jJQM

**Comment by Lombaxtard** [comment]
> well it was there ever since i added the feature so...

I guess it's the fact that I haven't played with CTV in fp for weeks, or I'm just being delusional



Regarding everything else I can only say that's it's fun seeing you <del>suffer</del> tinker with this kind of stuff
Well, believe me, in Draconic's implementation it didn't look too robust either

**Comment by relaxtakenotes** [comment]
pushed an update with that thing. set cl_nte_hack_drc to 1 and restart the map. also i did some logic changes so that head prediction is not as jittery.

**Comment by Lombaxtard** [comment]
is this hacks??? is vac gonna ban me now??
I installed new version, got into a map, set drc to 1, restarted the map, I can see that drc is still at 1 but no leggss
I disabled all addons beforehand to test this btw

**Comment by relaxtakenotes** [comment]
> is this hacks??? is vac gonna ban me now?? I installed new version, got into a map, set drc to 1, restarted the map, I can see that drc is still at 1 but no leggss I disabled all addons beforehand to test this btw

wow unfortunate

**Comment by Lombaxtard** [comment]
Hey, I was thinking about this for a while now...
There's this addon called <a href="https://steamcommunity.com/sharedfiles/filedetails/?id=108168164">Customisable Thirdperson</a> and it has a veeery unique feature - camera nodes. Basically CTP allows you to place these nodes and they will act as static cameras with proximity activation that will follow player (just like GMod "Camera" tool placed with RMB) until trace is interrupted by collision/you are in radius of another, closer node, as shown on this illustration here:
![image](https://github.com/relaxtakenotes/cool-thirdperson-view/assets/75931669/87f87816-1edd-4602-9a61-20667ad28a4c)
Anyway, the feature is pretty cool and is practically the only way to have early Resident Evil/Silent Hill-styled camera perspective in Garry's Mod (<a href="https://steamcommunity.com/sharedfiles/filedetails/?id=2792388721">PS1 Camera Style</a> is cool too, but it's very random, is relatively uncontrollable, and is impossible to use for real gameplay in tight spaces), but it has a serious downside - performance. Placing a bunch of nodes (and re-engaging CTP a couple of times) causes an ungodly amoung of lag which increases the more you toggle CTP on and off.

I just thought I'd share this idea with you in case you'd be willing to, perhaps, make your own version of this feature

**Comment by relaxtakenotes** [comment]
pretty interesting

**Comment by Lombaxtard** [comment]
Hey, I think you really should try out CTV with newly released <a href="https://steamcommunity.com/sharedfiles/filedetails/?id=3050234499">Dynamic Height 2</a>, it really adds up a lot, especially combined with hybrid firstperson (it basically fixes and sort of unifies CTV fp and actual fp view offsets, except for X and Y axes)
Here's a <a href="https://www.youtube.com/watch?v=MX3gcdmmH8M">video</a> that compares default view with dynamic height 2 (including with CTV on and off)
Dunno if this can be addressed, but there's also seems to be a "bug" at the end of the video at 3:38 that shows some jitter when using CTV thirdperson with DH2 enabled, just saying

**Comment by relaxtakenotes** [comment]
gah that jittering problem just cant go away, can it? i'll see what i can do.

**Comment by Lombaxtard** [comment]
> yeah well at this point i blame them and not my mod. i'm not doing anything wrong lol

According to <a href="https://staging.streamable.com/mxccil">this...</a>
...I'm not so sure
Anywho, I think that coolness of the Cool Thirdperson View outweighs this long going jiggly-wiggly nuisance

**Comment by relaxtakenotes** [comment]
pac3 can eat my ass and i'm not doing anything wrong. end of discussion

**Comment by Lombaxtard** [comment]
Too dumb to figure out pull requests, so:

concommand.Add("cl_ehw_clear_list", function(ply, cmd, args, arg_str)

if file.Exists("ehw_allowed.json", "DATA") then 
	file.Delete("ehw_allowed.json", "DATA")
end

	print("Whitelist cleared.")
end)

in here 
![image](https://github.com/relaxtakenotes/cool-thirdperson-view/assets/75931669/9ab0b5b4-d9c8-4d9c-9966-49e6ee13ccc5)
would be nice
